### ISSUE TITLE
Regarding Soft Prompts

### ISSUE THREAD

**Issue by ayushp88** [Varun0157/gpt-fine-tuning#1]
Hi, thank you for providing the code. Have you tested the trained soft prompts on a new paragraph to generate a summary? I am currently doing this, but the results are not as expected. The generated summaries are not close to the desired output.

**Comment by Varun0157** [comment]
Hi!

Try increasing the `max_length` parameter [in the dataloader](https://github.com/Varun0157/gpt-fine-tuning/blob/f1425f93d04d629bbde56314740521eeeffcfc3c/src/data.py#L28).

Currently, it is set at 256 due to limited compute. The reason I suspect this could be leading to unexpected results is because 256 tokens may not be sufficient to wholly capture either the input paragraph or the output summary, leading to a paraphrase being generated instead of a summary (as I mentioned in the report if you're interested). 

Let me know if you're able to do this, otherwise, I can try performing the fine tuning and providing you the weights. 

Thanks for giving it a try!
Varun

**Comment by ayushp88** [comment]
> Hi!
> 
> Try increasing the `max_length` parameter [in the dataloader](https://github.com/Varun0157/gpt-fine-tuning/blob/f1425f93d04d629bbde56314740521eeeffcfc3c/src/data.py#L28).
> 
> Currently, it is set at 256 due to limited compute. The reason I suspect this could be leading to unexpected results is because 256 tokens may not be sufficient to wholly capture either the input paragraph or the output summary, leading to a paraphrase being generated instead of a summary (as I mentioned in the report if you're interested).
> 
> Let me know if you're able to do this, otherwise, I can try performing the fine tuning and providing you the weights.
> 
> Thanks for giving it a try! Varun

Ok thanks! But I think, if I am giving it a paragraph of tokens less than 256 to summarize, then it should do well and give the correct summary. What do you think?

**Comment by Varun0157** [comment]
I think your hypothesis will be true if the data you are training it on also aligns with that, i.e., if you train it on paragraphs of < 256 tokens and their corresponding summaries. 
The data that I've used for the current weights I've shared has been trained on a corpus with larger paragraph inputs, so a large number of samples probably consist of paragraphs and summaries of max token length, potentially leading to odd results. Does that make sense? 

This is just a theory from my end though, I should train it with a larger max length, but I lack the compute at the moment.

**Comment by ayushp88** [comment]
> I think your hypothesis will be true if the data you are training it on also aligns with that, i.e., if you train it on paragraphs of < 256 tokens and their corresponding summaries. The data that I've used for the current weights I've shared has been trained on a corpus with larger paragraph inputs, so a large number of samples probably consist of paragraphs and summaries of max token length, potentially leading to odd results. Does that make sense?
> 
> This is just a theory from my end though, I should train it with a larger max length, but I lack the compute at the moment.

Ok got it! The average token count of the articles in CNN daily mail dataset is 781, should I try to train with max length of 781?

**Comment by Varun0157** [comment]
Yes that would be great! Do let me know how it turns out!

**Comment by ayushp88** [comment]
I did the training with max length as 781. You can find the results here [Soft_Prompts ](https://docs.google.com/document/d/1jAyVMBrCmx0sMrLKfFDP2AeBIVMow6x4iHFwECuyTYk/edit?usp=sharing). Should I also share the testing code here?

**Comment by Varun0157** [comment]
Wow, that looks great! 

Sure, I would love to see the testing code. If you want, you can make a PR (perhaps add the code to a file called `inference-sample.py` and I can merge it into the repository - that way you get the credit. 

If you would like credit for the results too, create a PR to add it to a file called `sample.txt` in `docs/`. 

If you're not free to do so, I can, so no pressure. 

Thanks again for trying it out!

**Comment by ayushp88** [comment]
Thank you, could you please have a look at the PR?

**Comment by ayushp88** [comment]
Hi, one more thing, when I trained the soft prompts with learning rate 0.0001, the generated summaries were better than when it was set to 0.0005. Have you also checked the rouge score; it is coming to be low. Also, I am currently working on fine-tuning for a classification task using prompt-based tuning. If you're interested, I would be happy to raise a pull request for this implementation as well.

**Comment by Varun0157** [comment]
Sure, if you're interested feel free to add it! 

I didn't do too much hyperparameter tuning on these implementations to be honest, the initial implementation was more for my understanding than for any practical purpose in particular. Thanks for the observation!

**Comment by ayushp88** [comment]
Ok, I will add it. Should I create a separate data.py file for classification task? I am thinking to perform it on open book question answering dataset.

**Comment by Varun0157** [comment]
If the rest of the code (the training and testing portions) are the same as the current code, then just add the Dataset to the current data.py and create a parameter in the get_dataloader function to select either CNNDailyMail or your classification dataset, I think that's cleaner.

**Comment by Varun0157** [comment]
Hi @ayushp88, I made a few changes to your code and tested it and it works! 
I've merged it into the repo. Thanks for your contribution. 

I'm closing this issue for the time-being.

**Comment by ayushp88** [comment]
Ok, thank you! I am working on the classification code.
